plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Currency.hs

```diff
@@ -122,11 +125,19 @@ forgeContract
     -> [(TokenName, Integer)]
     -> Contract s e Currency
 forgeContract pk amounts = do
-    refTxIn <- PK.pubKeyContract pk (Ada.lovelaceValueOf 1)
-    let theCurrency = mkCurrency (txInRef refTxIn) amounts
-        forgedVal   = forgedValue theCurrency
+    (txOutRef, txOutTx, pkInst) <- PK.pubKeyContract pk (Ada.lovelaceValueOf 1)
+    let theCurrency = mkCurrency txOutRef amounts
         curVali     = curPolicy theCurrency
-
-    let forgeTx = mustSpendInput refTxIn <> forgeValue forgedVal (Set.singleton curVali)
-    _ <- submitTx forgeTx
+        mpsHash     = monetaryPolicyHash curVali
+        lookups     = ScriptLookups 
+                        { slMPS = Map.singleton mpsHash curVali
+                        , slTxOutputs = Map.singleton txOutRef txOutTx
+                        , slOtherScripts = Map.singleton (Scripts.scriptAddress pkInst) (Scripts.validatorScript pkInst)
+                        , slOtherData = Map.empty
+                        , slScriptInstance = Nothing
+                        , slOwnPubkey = Nothing
+                        }
+    let forgeTx = Constraints.mustSpendScriptOutput txOutRef unitRedeemer
+                    <> foldMap (\(tn, i) -> Constraints.mustForgeValue mpsHash tn i) amounts
```

Comment:
No particular reason, yes it would be easier to use `Value` in `mustForgeValue`

Reply:
Actually there is a reason for representing the constraint as `MustForgeValue MonetaryPolicyHash TokenName Integer`. It makes it easier to check against the `valueForged` field of the `PendingTx` esp. if we have multiple `MustForgeValue` constraints in one `TxConstraints` value.

If the value forged is `vl` and we get a `MustForgeValue mps tn i` constraint we can just check that `vl[mps][tn] == i`, note the equals sign. To check a `MustForgeValue vl'` we could either do `vl == vl'` (which would break if there was another `MustForgeValue` constraint with a different value, even for a totally unrelated currency), or we could do `vl ``leq`` vl'` (which would also accept a forged amount that was greater than what we put in the constraint, which should be avoided IMO).

But I can add a smart constructor that takes a `Value` and produces the constraints for all of the entries in the `Value`.

Reply:
Or maybe we just want a way to project a `Value` down onto a `CurrencyId`? So you can do `v `along` curId == mustForge`?